rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -381,66 +382,177 @@ typedef struct CustomClientResponse
+      tempMsg.length = change->serializedPayload.length;
+      tempMsg.max_size = change->serializedPayload.max_size;
+      tempMsg.buffer = change->serializedPayload.data;
+      // memcpy(tempMsg.buffer, change->serializedPayload.data, tempMsg.length);
+      if (proxyData.readFromCDRMessage(&tempMsg)) {
+        // TODO(wjwwood): remove this logic and replace with check for
+        // a single partition which is prefixed with the ROS specific
+        // prefix.
+        auto partition_str = std::string("");
+        // don't use std::accumulate - schlemiel O(n2)
+        for (const auto & partition : proxyData.m_qos.m_partition.getNames()) {
+          partition_str += partition;
+        }
+        auto fqdn = partition_str + "/" + proxyData.topicName();
 
-    WriterProxyData proxyData;
-    CDRMessage_t tempMsg;
-    tempMsg.msg_endian = change->serializedPayload.encapsulation ==
-      PL_CDR_BE ? BIGEND : LITTLEEND;
-    tempMsg.length = change->serializedPayload.length;
-    memcpy(tempMsg.buffer, change->serializedPayload.data, tempMsg.length);
-    if (proxyData.readFromCDRMessage(&tempMsg)) {
-      bool mapModified = false;
-      // TODO(wjwwood): remove this logic and replace with check for
-      // a single partition which is prefixed with the ROS specific
-      // prefix.
-      auto partition_str = std::string("");
-      // don't use std::accumulate - schlemiel O(n2)
-      for (const auto & partition : proxyData.m_qos.m_partition.getNames()) {
-        partition_str += partition;
-      }
-      mapmutex.lock();
-      auto fqdn = partition_str + "/" + proxyData.topicName();
-      auto it = topicNtypes.find(fqdn);
-      if (change->kind == ALIVE) {
+        mapmutex.lock();
+        auto it = topicNtypes.find(fqdn);
         if (
           it == topicNtypes.end() ||
           it->second.find(proxyData.typeName()) == it->second.end())
         {
           topicNtypes[fqdn].insert(proxyData.typeName());
-          mapModified = true;
+          rmw_ret_t ret = rmw_trigger_guard_condition(graph_guard_condition_);
+          if (ret != RMW_RET_OK) {
+            fprintf(stderr, "failed to trigger graph guard condition: %s\n",
+              rmw_get_error_string_safe());
+          }
         }
-      } else {
+        mapmutex.unlock();
+      }
+    } else {
+      GUID_t readerGuid;
+      iHandle2GUID(readerGuid, change->instanceHandle);
+
+      if (participant_->get_remote_reader_info(readerGuid, proxyData)) {
+        // TODO(wjwwood): remove this logic and replace with check for
+        // a single partition which is prefixed with the ROS specific
+        // prefix.
+        auto partition_str = std::string("");
+        // don't use std::accumulate - schlemiel O(n2)
+        for (const auto & partition : proxyData.m_qos.m_partition.getNames()) {
+          partition_str += partition;
+        }
+        partition_str = "rt";  // TODO(dirk-thomas): we don't have the partition info here
```

Comment:
Since eProsima/FastRTPS@63ea163ddade9576cbf41c829c2494b30192925f ```get_remote_reader_info``` and ```get_remote_writer_info`` return the partition information.